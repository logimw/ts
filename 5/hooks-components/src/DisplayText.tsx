import React, {FC, useEffect, useState} from 'react';
import UserTodos from "./UserTodos";

interface DisplayTextProps {
    getUserFullName: (username: string) => Promise<string>;
}

const DisplayText: FC<DisplayTextProps> = ({getUserFullName}) => {
    const [txt, setTxt] = useState("");
    const [msg, setMsg] = useState("");
    const [todoControl, setTodoControl] = useState<ReturnType<typeof UserTodos>>();
    const onChangeTxt = (e: React.ChangeEvent<HTMLInputElement>) => {
        setTxt(e.target.value);
    }

    const onClickShowMsg = async (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
        e.preventDefault();
        setMsg(`Testowanie reacta, ${await getUserFullName(txt)}`);
        setTodoControl(<UserTodos username={txt} />);
    }


    return (
        <form>
            <div>
                <label htmlFor="">Your name</label>
            </div>
            <div>
                <input type="text" data-testid="user-input"  value={txt} onChange={onChangeTxt} />
            </div>
            <div>
                <button data-testid="input-submit" onClick={onClickShowMsg}>Show message</button>
            </div>
            <div>
                <label htmlFor="" data-testid="final-msg">{msg}</label>
            </div>
            <ul>{todoControl}</ul>
        </form>
    );
};

export default DisplayText;
