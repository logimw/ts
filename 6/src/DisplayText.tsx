import React, { useState, FC } from "react";
import UserTodos from "./UserTodos";

interface DisplayTextProps {
    getUserFullname: (username: string) => Promise<string>;
}

const DisplayText: FC<DisplayTextProps> = ({ getUserFullname }) => {
    const [txt, setTxt] = useState("");
    const [msg, setMsg] = useState("");
    const [todoControl, setTodoControl] = useState<ReturnType<typeof UserTodos>>();

    const onChangeTxt = (e: React.ChangeEvent<HTMLInputElement>) => {
        setTxt(e.target.value);
    }

    const onClickShowMsg = async (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
        e.preventDefault();
        setMsg(`Witaj na zajęciach z testowania Reacta, ${await getUserFullname(txt)}!`);
        //setTodoControl(<UserTodos username={txt} />);
    }

    /*
    const setUsersTodos = async () => {
        const usersResponse = await fetch('https://jsonplaceholder.typicode.com/users');
        if(usersResponse.ok) {
            const users = await usersResponse.json();
            const userByName = users.find((usr: any) => {
                return usr.username.toLowerCase() === txt;
            });
            console.log("użytkownik pobrany po nazwie", userByName);

            const todosResponse = await
                fetch('https://jsonplaceholder.typicode.com/todos');
            if(todosResponse.ok) {
                const todos = await todosResponse.json();
                const usersTodos = todos.filter((todo:any) => {
                    return todo.userId === userByName.id;
                });
                const todoList = usersTodos.map((todo:any) => {
                    return <li key={todo.id}>
                        {todo.title}
                    </li>
                });
                setTodos(todoList);
                console.log("lista zadań użytkownika", usersTodos);
            }
        }
    }
    //*/

    return (
        <form>
            <div>
                <label>Jak masz na imię?</label>
            </div>
            <div>
                <input data-testid="user-input"
                    value={txt} onChange={onChangeTxt} />
            </div>
            <div>
                <button data-testid="input-submit"
                    onClick={onClickShowMsg}>Wyświetl komunikat</button>
            </div>
            <div>
                <label data-testid="final-msg">{msg}</label> 
            </div>
            <ul style={{marginTop: '1rem', listStyleType:'none'}}>
                {todoControl}
            </ul>
        </form>
    )
}
export default DisplayText; 