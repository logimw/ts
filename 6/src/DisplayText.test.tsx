import React from "react";
import { render, fireEvent, cleanup, wait } from "@testing-library/react";
import DisplayText from "./DisplayText";
import "@testing-library/jest-dom/extend-expect";

jest.mock("./UserTodos");

describe("Test komponentu DisplayText", () => {
  const userFullName = "Janek Tester";

  const getUserFullnameMock = (
    username: string
  ): [Promise<string>, jest.Mock<Promise<string>, [string]>] => {
    const promise = new Promise<string>((res, rej) => {
      res(userFullName);
    });
    const getUserFullname = jest.fn(
      async (username: string): Promise<string> => {
        return promise;
      }
    );

    return [promise, getUserFullname];
  };

  it("jest wyświetlany bez problemów", () => {
    const username = "testuser";
    const [promise, getUserFullname] = getUserFullnameMock(username);

    const { baseElement } = render(<DisplayText getUserFullname={getUserFullname} />);
    expect(baseElement).toBeInTheDocument();
    //console.log(baseElement.innerHTML);
  });

  it("jest zgodny z migawką", () => {
    const username = "testuser";
    const [promise, getUserFullname] = getUserFullnameMock(username);

    const { baseElement } = render(<DisplayText getUserFullname={getUserFullname} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("uzyskuje tekst z pola", () => {
    const username = "testuser"; 
    const [promise, getUserFullname] = getUserFullnameMock(username);
    
    const { getByTestId } = render(<DisplayText getUserFullname={getUserFullname} />);
    const input = getByTestId("user-input");
    fireEvent.change(input, { target: { value:username } });
    expect(input).toBeInTheDocument();
    expect(input).toHaveValue(username);
  });

  it("wyświetla komunikat powitalny", async () => {
    const username = "testuser";
    const [promise, getUserFullname] = getUserFullnameMock(username);


    const msg = `Witaj na zajęciach z testowania Reacta, ${userFullName}!`;
    const { getByTestId } = render(<DisplayText getUserFullname={getUserFullname} />);
    const input = getByTestId("user-input");
    const label = getByTestId("final-msg");
    fireEvent.change(input, { target: { value: username } });
    const btn = getByTestId("input-submit");
    fireEvent.click(btn);

    expect(label).toBeInTheDocument();
    await wait(() => promise);
    expect(label.innerHTML).toBe(msg);
  });

});
