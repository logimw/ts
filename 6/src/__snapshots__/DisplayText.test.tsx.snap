// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test komponentu DisplayText jest zgodny z migawką 1`] = `
<body>
  <div>
    <form>
      <div>
        <label>
          Jak masz na imię?
        </label>
      </div>
      <div>
        <input
          data-testid="user-input"
          value=""
        />
      </div>
      <div>
        <button
          data-testid="input-submit"
        >
          Wyświetl komunikat
        </button>
      </div>
      <div>
        <label
          data-testid="final-msg"
        />
      </div>
      <ul
        style="margin-top: 1rem; list-style-type: none;"
      />
    </form>
  </div>
</body>
`;
